File: src/algorithms/matrices/construction/pascals-triangle/sources/pascals-triangle_test.py
import importlib

pascals_triangle_mod = importlib.import_module("pascals-triangle")
pascals_triangle = pascals_triangle_mod.pascals_triangle


def test_returns_single_row_for_num_rows_1():
    assert pascals_triangle(1) == [[1]]


def test_returns_correct_triangle_for_num_rows_2():
    assert pascals_triangle(2) == [[1], [1, 1]]


def test_returns_correct_triangle_for_num_rows_3():
    assert pascals_triangle(3) == [[1], [1, 1], [1, 2, 1]]


def test_returns_correct_triangle_for_num_rows_5():
    assert pascals_triangle(5) == [[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1]]


def test_returns_correct_triangle_for_num_rows_6():
    assert pascals_triangle(6) == [
        [1],
        [1, 1],
        [1, 2, 1],
        [1, 3, 3, 1],
        [1, 4, 6, 4, 1],
        [1, 5, 10, 10, 5, 1],
    ]


def test_inner_cell_is_sum_of_two_above():
    result = pascals_triangle(5)
    for row_idx in range(2, len(result)):
        current_row = result[row_idx]
        above_row = result[row_idx - 1]
        for col_idx in range(1, len(current_row) - 1):
            assert current_row[col_idx] == above_row[col_idx - 1] + above_row[col_idx]


def test_all_edge_cells_are_1():
    result = pascals_triangle(6)
    for row in result:
        assert row[0] == 1
        assert row[-1] == 1


def test_row_length_equals_row_index_plus_one():
    result = pascals_triangle(5)
    for row_idx, row in enumerate(result):
        assert len(row) == row_idx + 1


def test_returns_empty_array_for_num_rows_0():
    assert pascals_triangle(0) == []


if __name__ == "__main__":
    test_returns_single_row_for_num_rows_1()
    test_returns_correct_triangle_for_num_rows_2()
    test_returns_correct_triangle_for_num_rows_3()
    test_returns_correct_triangle_for_num_rows_5()
    test_returns_correct_triangle_for_num_rows_6()
    test_inner_cell_is_sum_of_two_above()
    test_all_edge_cells_are_1()
    test_row_length_equals_row_index_plus_one()
    test_returns_empty_array_for_num_rows_0()
    print("All tests passed!")

File: src/algorithms/matrices/construction/pascals-triangle/sources/PascalsTriangle_test.cpp
// g++ -std=c++17 -o pascals_triangle_test PascalsTriangle_test.cpp && ./pascals_triangle_test
#include "PascalsTriangle.cpp"
#include <cassert>
#include <iostream>

int main() {
    // test: returns [[1]] for numRows=1
    {
        auto result = pascalsTriangle(1);
        assert(result.size() == 1);
        assert((result[0] == std::vector<int>{1}));
    }

    // test: returns correct triangle for numRows=2
    {
        auto result = pascalsTriangle(2);
        assert(result.size() == 2);
        assert((result[0] == std::vector<int>{1}));
        assert((result[1] == std::vector<int>{1, 1}));
    }

    // test: returns correct triangle for numRows=3
    {
        auto result = pascalsTriangle(3);
        assert(result.size() == 3);
        assert((result[2] == std::vector<int>{1, 2, 1}));
    }

    // test: returns correct triangle for numRows=5
    {
        auto result = pascalsTriangle(5);
        assert(result.size() == 5);
        assert((result[3] == std::vector<int>{1, 3, 3, 1}));
        assert((result[4] == std::vector<int>{1, 4, 6, 4, 1}));
    }

    // test: returns correct triangle for numRows=6
    {
        auto result = pascalsTriangle(6);
        assert(result.size() == 6);
        assert((result[5] == std::vector<int>{1, 5, 10, 10, 5, 1}));
    }

    // test: each inner cell is the sum of the two cells above
    {
        auto result = pascalsTriangle(5);
        for (size_t rowIdx = 2; rowIdx < result.size(); rowIdx++) {
            const auto& currentRow = result[rowIdx];
            const auto& aboveRow = result[rowIdx - 1];
            for (size_t colIdx = 1; colIdx < currentRow.size() - 1; colIdx++) {
                assert(currentRow[colIdx] == aboveRow[colIdx - 1] + aboveRow[colIdx]);
            }
        }
    }

    // test: all edge cells are 1
    {
        auto result = pascalsTriangle(6);
        for (const auto& row : result) {
            assert(row.front() == 1);
            assert(row.back() == 1);
        }
    }

    // test: row at index rowIdx has rowIdx+1 elements
    {
        auto result = pascalsTriangle(5);
        for (size_t rowIdx = 0; rowIdx < result.size(); rowIdx++) {
            assert(result[rowIdx].size() == rowIdx + 1);
        }
    }

    // test: returns empty for numRows=0
    {
        auto result = pascalsTriangle(0);
        assert(result.empty());
    }

    std::cout << "All tests passed!" << std::endl;
    return 0;
}
